ports.js

```diff
@@ -1,51 +1,228 @@
+  return Promise.resolve(null);
+}
+
+function dropPassphrase() {
+  sessionStorage.removeItem("temporaryPassphrase");
+  return Promise.resolve(null);
 }
 
-app.ports.decryptData.subscribe(function(data) {
-  console.log(data);
-  if (!data.content) {
-    app.ports.dataDecrypted.send(null);
-    return;
+function getPassphrase() {
+  // Insecurely retrieving the passphrase.
+  const passphrase = sessionStorage.getItem("temporaryPassphrase");
+  if (passphrase === null) {
+    return null;
   }
-  decrypt(data.passphrase, data.content)
-    .then(content => {
-      app.ports.dataDecrypted.send(content);
-    })
-    .catch(err => {
-      console.error('Error decrypting', err);
-      app.ports.dataNotDecrypted.send(err.message);
+  return Promise.resolve(atob(passphrase));
+}
+
+function getItem(key) {
+  if (!IS_WEB_EXTENSION) {
+    return Promise.resolve(localStorage.getItem(key));
+  } else {
+    return new Promise(function(resolve, reject) {
+      chrome.storage.local.get(key, function(data) {
+        resolve(data[key] || null);
+      });
     });
-});
+  }
+}
 
-app.ports.encryptData.subscribe(function(data) {
-  if (!data.content) {
-    return;
+function setItem(key, value) {
+  if (!IS_WEB_EXTENSION) {
+    if (value === null) {
+      localStorage.removeItem(key);
+    } else {
+      localStorage.setItem(key, value);
+    }
+    return Promise.resolve(null);
+  } else {
+    return new Promise(function(resolve, reject) {
+      let payload = {};
+      payload[key] = value;
+      chrome.storage.local.set(payload, () => {
+        resolve(null);
+      });
+    });
+  }
+}
+
+
+function createElmApp(flags) {
+  let app;
+  if (typeof Elm === 'undefined') {
+    const Elm = require('./hoverpad.js');
+    app = Elm.Main.fullscreen(flags);
+  } else {
+    app = Elm.Main.fullscreen(flags);
   }
-  encrypt(data.passphrase, data.content.replace(/<br>$/g, ''))
-    .then(encryptedContent => {
-      app.ports.dataEncrypted.send(encryptedContent);
-    })
-    .catch(err => {
-      console.error('Error encrypting', err);
-      app.ports.dataNotEncrypted.send(err.message);
+
+  app.ports.getData.subscribe(function() {
+    getItem(CONTENT_KEY)
+      .then(function(data) {
+        app.ports.newData.send([CONTENT_KEY, data || ""]);
+      })
+      .catch(function(err) {
+        console.error(err);
+        app.ports.newError.send('Nothing retrieved: ' + err.message);
+      });
+  });
+
+  app.ports.saveData.subscribe(function(data) {
+    setItem(data.key, data.content)
+      .then(function() {
+        app.ports.dataSaved.send(data.key);
+      })
+      .catch(function(err) {
+        console.error(err);
+        app.ports.newError.send('Nothing retrieved: ' + err.message);
+      });
+  });
+
+  app.ports.savePassphrase.subscribe(function(passphrase) {
+    storePassphrase(passphrase)
+      .then(function() {
+        console.log("Passphrase saved");
+      })
+      .catch(function(err) {
+        console.error(err);
```

Comment:
We may want to disable console logging by default (enable it for debug builds or smthg), so users won't have these messages in their browser console. THis may be addressed in a followup PR though.